### ISSUE TITLE
[prefer-const] false positive when `parserOptions.ecmaVersion` is `"latest"`

### ISSUE THREAD

**Issue by strangedev** [typescript-eslint/typescript-eslint#3793]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
    "env": {
        "browser": true,
        "es2021": true
    },
    "extends": [
        "eslint:recommended",
        "plugin:@typescript-eslint/recommended"
    ],
    "parserOptions": {
        "ecmaVersion": "latest",
        "sourceType": "module"
    },
    "rules": {
        "prefer-const": "error"
    }
}
```

```TS
const foobar = function (): void {
  let bar;

  try {
    bar = 3;
  } catch {
    console.log('err');
  }

  console.log(bar);
};

export {
  foobar
};
```

**Expected Result**

No error is reported for the `prefer-const` rule, as declaration and assignment of `bar` are not in the same scope.

**Actual Result**

The following error is reported: `error  'bar' is never reassigned. Use 'const' instead  prefer-const`.

**Additional Info**

[I filed this bug with eslint before](https://github.com/eslint/eslint/issues/14977) since it is their rule.
As pointed out in the issue
> The rule checks if the assignment is in the same scope [here](https://github.com/eslint/eslint/blob/db1518374a5e88efedf1ed4609d879f3091af74f/lib/rules/prefer-const.js#L245), by using the scope analysis info, which is in this case provided by @typescript-eslint/parser.

Interestingly, this false positive is only reported when I use `@typescript-eslint/parser` (hence I am opening an issue here) __and__ `parserOptions.ecmaVersion` is set to `"latest"`. Once `parserOptions.ecmaVersion` is set to `12`, the rule doesn't report any error anymore.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.29.3` |
| `@typescript-eslint/parser`        | `4.29.1` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.32.0` |
| `node`                             | `16.7.0` |
